### PR TITLE
General changes on pipeline.py

### PR DESCRIPTION
**Pull request by rodrigomoliveira1**
Some general modifications to `pipeline`, which does not affect its functioning.
These PR is related to code reading and documentation

- add a brief module description;
- move the example at the end to Pipeline docstring (there's still work to do on it);
- classes' names changed to receive capital letters;
- plot methods' names changed to receive "plot_" prefix;
- general translations from portuguese to english;
- some documentation improvements (there's still work to do on it);
- changes on arguments to simplify the user inputs: for example: instead of the need to write `hyphotesis_test(kind="Komolgorov-Smirnov")`, the user inputs only `hyphotesis_test(kind="ks")`;
- run Black and Isort.

- Neural Network models, figures and tables are saved in separated folders inside ROSS-ML package. To load them, it's only required to pass the folder's name (no need to insert the entire path).

### CODE REVIEW

File: Pipeline/pipeline.py [hunk 1]
```diff
@@ -73,100 +108,99 @@ def __init__(self, df):
...
-        Array
-             y_test - labels destined for test.
+        X_train : array
+            Features destined for training.
+        X_test : array
+            Features destined for test.
+        y_train : array
+            Labels destined for training.
+        y_test : array
+             Labels destined for test.
```
Comment (raphaeltimbo) [inline Pipeline/pipeline.py:203]: In the validation method we have changed from X to x.
From what I understand tensorflow standard is to use `x_train`, `y_train`... while scikit-learn uses `X_train`, `y_train`...
Since we have already change part of the code from X to x, I think we should also change it here, following the tensorflow standard of `x_train` which is also pep8 compliant.

File: Pipeline/pipeline.py [hunk 2]
```diff
@@ -367,20 +418,45 @@ def hypothesis_test(self, kind="Komolgorov-Smirnov", p_value=0.05):
...
+        y : TYPE
+            DESCRIPTION.
+
+        Returns
+        -------
+        train : pd.DataFrame
+            DESCRIPTION.
+        test : pd.DataFrame
+            DESCRIPTION.
+
```
Comment (raphaeltimbo) [inline Pipeline/pipeline.py:437]: Improve this part of the docs given information about the method and its parameters.

File: Pipeline/pipeline.py [hunk 3]
```diff
@@ -533,22 +608,25 @@ def q_q_plot(self):
             )
             plt.close()
 
-    def standardized_error_plot(self):
-        var = list(self.train.columns)  
-        erro = self.test - self.train
-        erro = erro / erro.std()
-        erro.dropna(inplace=True)
+    def plot_standardized_error(self):
+        """Plot and save the graphic for standardized error.
```
Comment (raphaeltimbo) [inline Pipeline/pipeline.py:612]: Plot and automatically save is not good practice. 
Here we should follow the same format that we have for our plots in the ROSS main package, returning the figure to the user. With matplotlib we should return the axes (ax object), something like:
```python
def plot(x, y, ax=None):
    if ax is None:
        ax = plt.gca()
    ax.plot(x, y)
    return ax
```

**Review by raphaeltimbo — CHANGES_REQUESTED** [review]

File: Pipeline/pipeline.py (hunk 2 above)
Comment (rodrigomoliveira1) [reply Pipeline/pipeline.py:437]: Yes, I still have to improve this and the others docstrings. I've just opened the style to complement it later.

File: Pipeline/pipeline.py [hunk 4]
```diff
@@ -184,11 +218,11 @@ def data_scaling(self, test_size, scaling=True, scalers=[]):
         else:
             self.scaler1 = None
             self.scaler2 = None
+
         return self.X_train, self.X_test, self.y_train, self.y_test
 
     def build_Sequential_ANN(self, hidden, neurons, dropout_layers=[], dropout=[]):
```
[the remark this answers is not in the document]
Comment (rodrigomoliveira1) [reply Pipeline/pipeline.py:224]: I had forgotten about changing it

File: Pipeline/pipeline.py [hunk 5]
```diff
@@ -62,7 +50,54 @@ def HTML_formater(df, name):
...
+    df : pd.Dataframe
+
+    Examples
+    --------
+    >>> import pandas as pd
+    >>> import rossml as rsml
+
+    Importing and collecting data
+    >>> df = pd.read_csv('seal_fake.csv')
+    >>> df_val= pd.read_csv('xllaby_data-componentes.csv')
```
[the remark this answers is not in the document]
Comment (rodrigomoliveira1) [reply Pipeline/pipeline.py:67]: I was thinking in add a new folder "example" to add some .csv files to use as example in docstring, etc, just like it's done in ROSS. Furthemore, we still have to do some changes in this example. It's too big and could take much time to run, when we have Travis and AppVeyor.

File: Pipeline/pipeline.py [hunk 6]
```diff
@@ -1,51 +1,39 @@
-# -*- coding: utf-8 -*-
-"""
-Created on Tue Jul 21 12:10:24 2020
+"""ROSS Machine Learning.
 
-@author: Luan
+ROSS-ML is a module to create neural networks aimed at calculating rotodynamic
+coefficients for bearings and seals.
```
[the remark this answers is not in the document]
Comment (LuanDaSilva) [reply Pipeline/pipeline.py:4]: 1) Divide data in two groups: labels and features. After dividing it, choose a sample to cross-validate the data.
2 - question 1) Please rephrase your question. What exactly do you mean by assumptions? 
2 - question 2) It depends. For classification problems, sigmoids are prefered on output. For regression, you should use ReLU, or any other. The layer will affect the topology to optmize the loss function
3) Cross-validation gives you a set to be compared with the trained one. In general, you should use a metric according to your problem. It varies a lot. For regression, you could use R^2,R^2 adjusted,MAE or MSE. For classification, cross-entropy is prefered. Here I also added KS test. The idea between KS test is that the original data and trained data should arise from the same distribution if the train succeded

File: Pipeline/pipeline.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (LuanDaSilva) [reply Pipeline/pipeline.py:67]: seal_fake.csv is a expanded dataset with 50k points and xllaby_data-components.csv is a dataset with 4k points from XLTRC

File: Pipeline/pipeline.py [hunk 7]
```diff
@@ -62,7 +50,54 @@ def HTML_formater(df, name):
...
+
+    Importing and collecting data
+    >>> df = pd.read_csv('seal_fake.csv')
+    >>> df_val= pd.read_csv('xllaby_data-componentes.csv')
+
+    >>> df_val.fillna(df.mean)
+    >>> D = Pipeline(df)
+    >>> D.set_features(0, 20)
+    >>> D.set_labels(20, len(D.df.columns))
+    >>> D.feature_reduction(15)
```
[the remark this answers is not in the document]
Comment (LuanDaSilva) [reply Pipeline/pipeline.py:73]: I will create a README to answer all this questions and new ones from pull requests.

File: Pipeline/pipeline.py [hunk 8]
```diff
@@ -62,7 +50,54 @@ def HTML_formater(df, name):
...
+    >>> D.data_scaling(0.1, scalers=[RobustScaler(), RobustScaler()], scaling=True)
+    >>> D.build_Sequential_ANN(4, [50, 50, 50, 50])
+    >>> model, predictions = D.model_run(batch_size=300, epochs=1000)
+
+    Get the model configurations to change it afterwards
+    >>> # model.get_config()
```
[the remark this answers is not in the document]
Comment (LuanDaSilva) [reply Pipeline/pipeline.py:79]: This model.get_config is a command to change neural networks parameters after adjusting it on pipeline without using it again. To further details, please check keras documentatation.

File: Pipeline/pipeline.py (hunk 3 above)
Comment (LuanDaSilva) [reply Pipeline/pipeline.py:612]: The figures are saved for a HTML file resuming all plots. Without it a pratical front-end isn't feasible, because it will be necessary to include a Java Script to make requests to .py file in order to generate the needed figures dynamically, which will be expensive to the browser. This HTML will be a resume of all results. For HTML or any front-end language, the best pratice is to create a folder with the needed images to generate a HTML page. We can also remove the front end, but I should not recommend it, because graphics itself could be misleading without proper context, which I created on HTML file to guide the user to correctly use pipeline and interpret the results.

File: Pipeline/pipeline.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (LuanDaSilva) [reply Pipeline/pipeline.py:4]: First question with more details:

"How are we supposed to train the neural network - with data from other models, CFD results or test data?"

- 
You can use any type of data as long as it is possible to convert it to a structured DataFrame form in which you can identify the  exogenous variables (inputs) and endogenous ones (output). The main idea of pipeline is scalability. To sum up: you must look to your problem and make the following questions: 1)"Can the problem  be converted into structured data?" 2)"Is it possible to divide it clearly in inputs and outputs?"

File: Pipeline/pipeline.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (LuanDaSilva) [reply Pipeline/pipeline.py:67]: "I thought we only needed one file with the data, and the code would take care of splitting that in training and validation data." This is done in Pipeline. It will be clear when I finish the README

File: Pipeline/pipeline.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (raphaeltimbo) [reply Pipeline/pipeline.py:4]: > Divide data in two groups: labels and features. After dividing it, choose a sample to cross-validate the data.
2 - question 1) Please rephrase your question. What exactly do you mean by assumptions?
2 - question 2) It depends. For classification problems, sigmoids are prefered on output. For regression, you should use ReLU, or any other. The layer will affect the topology to optmize the loss function

I mean that, as a user, I would like to know the basics of how the code inside the 'Pipeline' class works, so layers type was just an example of information that I would like to see here at the documentation. 

> Cross-validation gives you a set to be compared with the trained one. In general, you should use a metric according to your problem. 

For our specific problem (taking the seal as an example), do we have any recommendation for how the user should carry out the validation?

File: Pipeline/pipeline.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (raphaeltimbo) [reply Pipeline/pipeline.py:67]: To me it is not clear how to generate the `seal_fake.csv` file. Maybe we should have this information on the documentation or maybe the expanded dataset could be generated automatically by the code without the user having to provide the file path (don't know if that is possible).

File: Pipeline/pipeline.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (raphaeltimbo) [reply Pipeline/pipeline.py:79]: But do we need to execute this command when running the example? 
Currently we have:
```python
>>> #model.get_config()
>>> D.model_history()
>>> D.metrics()
```
Should the code be:
```python
>>> model.get_config()
>>> D.model_history()
>>> D.metrics()
```
?

If the `model.get_config()` is not necessary to run the example I would remove this line.

File: Pipeline/pipeline.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (LuanDaSilva) [reply Pipeline/pipeline.py:4]: Yes, we have. The first metric used in validation is R squared. when it is close to one, the odds of obtaining a good result are high. The R^2 must be used between the chunk of data used to cross validation and the trained data. In a ideal world, trained data will be equal tested data and, as a result, R^2 = 1, which tells us that tested data = trained data. When you avail the problem by the means of R^2, you are asking to your model how close the trained set are closed to the line tested data = trained data. This kind of result are also shown in a q-q plot graph

[the remark this answers is not in the document]
Comment (LuanDaSilva) [reply Pipeline/pipeline.py:4]: If you want a notion of distance, you should use the Mean Squared Error, Mean Absolute error or anything with the same concept. The main idea here is to measure how far your points are from its refferences in average. MSE punishes hard the furthest points, while its very soft to the closest ones. Note that your scale is important: use MSE on data with order of 1e-1 it is useless. MAE have the same idea as MSE, but treat only the distance between the points in modulus, instead of punishing the furthest ones by squaring it all the points. The metric used will depend strongly on which typing of problem you are facing with

File: Pipeline/pipeline.py (hunk 3 above)
Comment (raphaeltimbo) [reply Pipeline/pipeline.py:612]: You can still save the figures for the HTML, it would just be in a different part of your code.
Something like:
```python
def generate_html():
    ...
    # save figures
    error_plot = self.plot_standardized_error()
    plt.savefig('error_plot.png')
```
Breaking this into separate parts of the code will allow the user to call the plot in other contexts in which they don't  want to save the plot.

File: Pipeline/pipeline.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (LuanDaSilva) [reply Pipeline/pipeline.py:4]: "I mean that, as a user, I would like to know the basics of how the code inside the 'Pipeline' class works, so layers type was just an example of information that I would like to see here at the documentation."

Fair enough. I will try to answer all these kind of question in a robust README, with all the required information to use properly pipeline without any major issues.

[the remark this answers is not in the document]
Comment (raphaeltimbo) [reply Pipeline/pipeline.py:4]: `Yes, we have. The first metric used in validation is R squared. when it is close to one, the odds of obtaining a good result are high. The R^2 must be used between the chunk of data used to cross validation and the trained data. In a ideal world, trained data will be equal tested data and, as a result, R^2 = 1, which tells us that tested data = trained data. When you avail the problem by the means of R^2, you are asking to your model how close the trained set are closed to the line tested data = trained data. This kind of result are also shown in a q-q plot graph`

Nice explanation! I was thinking of having something like this in our docs, always focusing on our problem type, which will always be something like the one we had for the seal: some input such as inlet_temperature, inlet_pressure, radius, etc. and output kxx, kxy, kyx, kyy,  cxx, cxy, cyx, cyy.

File: Pipeline/pipeline.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (LuanDaSilva) [reply Pipeline/pipeline.py:79]: No,no...This model.get_config() is a advance tool. Is it used as ,say,a "last resort"

[the remark this answers is not in the document]
Comment (LuanDaSilva) [reply Pipeline/pipeline.py:79]: One example of use is: imagine that you have just discovered that the first activation layer of your problem must be a 'tanh' and you have chosen a 'sigmoid'. To solve this situation, you use model.get_config to access your model and to change the activation function to the proper one. This could also be done with dropout, regularization and more.

File: Pipeline/pipeline.py [hunk 9]
```diff
@@ -41,11 +58,11 @@ def HTML_formater(df, name):
                    </body>
                     </html>.
                """
-    with open("tables/{}.html".format(name), "w") as f:
+    with open(path / "tables/{}.html".format(name), "w") as f:
         f.write(html_string.format(table=df.to_html(classes="mystyle")))
 
 
-class Pipeline:
+class Pipeline(object):
```
Comment (raphaeltimbo) [inline Pipeline/pipeline.py:65]: In python 3 we do not need to inherit from object.

File: Pipeline/pipeline.py [hunk 10]
```diff
@@ -94,12 +111,13 @@ class Pipeline:
...
 
     def __init__(self, df):
-        path1 = "img"
-        path2 = "tables"
-        if not os.path.isdir(path1):
-            os.makedirs(path1)
-        if not os.path.isdir(path2):
-            os.makedirs(path2)
+        path1 = Path(__file__).parent / "img"
+        path2 = Path(__file__).parent / "tables"
```
Comment (raphaeltimbo) [inline Pipeline/pipeline.py:115]: I suggest using more explicit names such as `path_img` instead of `path_1`.

File: Pipeline/pipeline.py [hunk 11]
```diff
@@ -173,36 +191,34 @@ def feature_reduction(self, n):
...
-        y_test : array
-             Labels destined for test.
+        Array
+            x_train - features destined for training.
+        Array
+             x_test -features destined for test.
+        Array
+            y_train - labels destined for training.
+        Array
+            y_test - labels destined for test.
```
Comment (raphaeltimbo) [inline Pipeline/pipeline.py]: Format was following the numpydoc, now it is not. This should be reverted.

File: Pipeline/pipeline.py [hunk 12]
```diff
@@ -173,36 +191,34 @@ def feature_reduction(self, n):
 
         return self.X
 
-    def data_scaling(self, test_size, scaling=True, scalers=None):
+    def data_scaling(self, test_size, scaling=True, scalers=[]):
```
Comment (raphaeltimbo) [inline Pipeline/pipeline.py:194]: It is not recommended using a list as a default parameter.

File: Pipeline/pipeline.py [hunk 13]
```diff
@@ -475,85 +495,123 @@ def validation(self, x, y):
...
+class Model(object):
     def __init__(self, name):
         self.name = name
 
     def load_model(self):
-        self.model = load_model(r"{}/{}.h5".format(self.name, self.name))
-        self.columns = load(
-            open(r"{}/{}_columns.pkl".format(self.name, self.name), "rb")
-        )
+        """Load a neural netowork model from rossml folder.
```
Comment (raphaeltimbo) [inline Pipeline/pipeline.py:539]: Typo netowork.

Comment (raphaeltimbo) [reply Pipeline/pipeline.py:539]: I also suggest using only `load` instead of `load_model`.

File: Pipeline/pipeline.py (hunk 9 above)
Comment (rodrigomoliveira1) [reply Pipeline/pipeline.py]: Fixed

File: Pipeline/pipeline.py (hunk 10 above)
Comment (rodrigomoliveira1) [reply Pipeline/pipeline.py]: Done!

File: Pipeline/pipeline.py (hunk 11 above)
Comment (rodrigomoliveira1) [reply Pipeline/pipeline.py]: Done!

File: Pipeline/pipeline.py (hunk 12 above)
Comment (rodrigomoliveira1) [reply Pipeline/pipeline.py]: Done!

File: Pipeline/pipeline.py (hunk 13 above)
Comment (rodrigomoliveira1) [reply Pipeline/pipeline.py]: Done!

**Review by raphaeltimbo — APPROVED** [review]
